spotbox/gui.py

```diff
@@ -0,0 +1,459 @@
+        for column in self._columns:
+            column.yview(*args)
+
+    def search(self, searchterm):
+        self._datasheet.search(searchterm)
+        self.refresh()
+        self._turn_all_arrows_off()
+
+
+
+class Header(tk.Frame, object):
+    """Header object. In charge of coordinating behavior of
+    all widgets in header frame.
+    """
+    def __init__(self, master, menus, lookuptable, playback_obj, config):
+        super(Header, self).__init__(master)
+        #newframe = tk.Frame(master)
+        #newframe.pack()
+        self.pack()
+        # the widgets that live in the header
+        buttonarray = LoadAndPlayButtons(self, menus, playback_obj, config.num_spots)
+        categoryselect = CategorySelect(self, lookuptable.order, menus, self)
+        self.searchbox = SearchBox(self, menus)
+
+        # add the graphic...
+        graphicfile = config.graphic
+        graphic = tk.PhotoImage(file=graphicfile)
+        graphicw = tk.Label(self, image=graphic)
+        graphicw.image = graphic
+
+        # organize and pack the insides, and then itself:
+        categoryselect.pack(side=tk.BOTTOM, fill=tk.X)
+        self.searchbox.pack(side=tk.BOTTOM, fill=tk.X)
+        buttonarray.pack(side=tk.RIGHT, fill=tk.BOTH)
+        graphicw.pack(side=tk.LEFT)
+
+        #graphicw.grid(row=0, column=0)
+        #buttonarray.grid(row=0, column=1)
+        #self.searchbox.grid(row=2, column=0)
+        #categoryselect.grid(row=2, column=1)
+
+    def update(self):
+        # currently, only cleanup is clearing search box:
+        self.searchbox.clear_text()
+
+
+class Menus(object):
+    """Menus object.
+
+    Contains references to all existing menus. An instance is
+    passed to other widgets, so that if menu manipulation is needed, it is
+    coordinated through this object.
+    """
+    def __init__(self, master, datasheets, config):
+        self._menus = {key: MenuOfSpots(master, self, config.headers[key],
+                                           datasheets[key])
+                          for key in config.order}
+        for menu in self._menus.values():
+            menu.refresh(reset_spots=True)
+
+        # Defaults to first in order.
+        self._currentmenu = self._menus[config.order[0]]
+        self._currentmenu.pack(expand=True, fill=tk.BOTH)
+        # the last file (full path) that the user clicked:
+        self._spot_to_load = None
+
+    def switch_to(self, newkey):
+        """Clears current menu, packs new menu,
+        and then fills the menu that was just cleared (in order that you
+        don't have to wait for menu to be randomized/etc.
+        """
+        self._currentmenu.pack_forget()
+        oldmenu = self._currentmenu
+        self._currentmenu = self._menus[newkey]
+        self._currentmenu.pack(expand=True, fill=tk.BOTH)
+        oldmenu.refresh(reset_spots=True)
+
+    def search(self, searchterm):
+        return self._currentmenu.search(searchterm)
+
+# class GUIState(object):
```

Comment:
Oh, this. Decide on whether a state object is smart or not. Having loadable spot in menu state isn't crazy.